app/src/test/java/com/android/foundamentals/workshop01/Workshop1PresenterTest.kt

```diff
@@ -0,0 +1,180 @@
+package com.android.foundamentals.workshop01
+
+import com.android.fundamentals.domain.login.LoginInteractor
+import com.android.fundamentals.workshop01.Workshop1Presenter
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.test.TestCoroutineDispatcher
+import kotlinx.coroutines.test.runBlockingTest
+import org.junit.After
+import org.junit.Assert
+import org.junit.Before
+import org.junit.Test
+
+@ExperimentalCoroutinesApi
+internal class Workshop1PresenterTest {
+
+    private lateinit var testDispatcher: TestCoroutineDispatcher
+
+    @Before
+    fun setUp() {
+        testDispatcher = TestCoroutineDispatcher()
+    }
+
+    @After
+    fun cleanUp() {
+        testDispatcher.cleanupTestCoroutines()
+    }
+
+    @Test
+    fun `assert login UserNameError`() = runBlockingTest {
+        // Prepare all components for testing
+        val (interactor, presenter, view) = createComponents()
+
+        // Call presenter method to attach view
+        // TODO 01: Uncomment this code
+        //presenter.attachView(view)
+
+        // Checking the view
+        view.assert(
+            expectedLoading = false,
+            expectedUserNameErrorShownTimes = 0,
+            expectedPasswordErrorShownTimes = 0,
+            expectedSuccessShownTimes = 0
+        )
+
+        // Call presenter method to attach view
+        // TODO 02: Uncomment this code and run the test
+        //presenter.login(userName = "", password = "")
+
+        // Checking the view
+        view.assert(
+            expectedLoading = true,
+            expectedUserNameErrorShownTimes = 0,
```

Comment:
expectedUserNameErrorShownTimes = TODO(),
             expectedPasswordErrorShownTimes = TODO(),
             expectedSuccessShownTimes = TODO(),
with comments?
and same for all cases below?